### DISCUSSION TITLE
菜单快捷键如何设置，目前重载窗口的菜单框架内有设置快捷键，但是不生效

### DISCUSSION THREAD

**Discussion by shilin8805** [opensumi/core#1757] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/3340210/194742504-236baa2e-ee82-4d1e-8005-3eb9f7cd7399.png)
![image](https://user-images.githubusercontent.com/3340210/194742518-ac328ad2-8189-4eb1-8035-f2274ae595ee.png)
如上图，框架有设置快捷键，但是不生效。

**Comment by bytemain** [comment]
请问 `registerMenuItem` 的代码是怎么写的呢？

---



我这里 2.20 测试是生效了的，不过因为 OpenSumi 内部（ide-electron-basic）已经注册了keybinding，想替换的话要先取消注册。

<img width="1021" alt="CleanShot 2022-10-09 at 15 43 00@2x" src="https://user-images.githubusercontent.com/13938334/194744334-2d680a25-3a17-4ebf-8b4d-c9de9e0ce191.png">

这个 Module 的位置：
```ts
import { ElectronBasicModule } from '@opensumi/ide-electron-basic/lib/browser';
```


```ts
import {
  Domain,
  KeybindingContribution,
  KeybindingRegistry,
} from '@opensumi/ide-core-browser';

import {
  MenuContribution,
  IMenuRegistry,
} from '@opensumi/ide-core-browser/lib/menu/next';

const COMMAND_ID = 'electron.reload';

@Domain(MenuContribution, KeybindingContribution)
export class MenubarContribution
  implements MenuContribution, KeybindingContribution
{
  registerMenus(registry: IMenuRegistry) {
    registry.registerMenuItem('windowMenuBar', {
      command: {
        id: COMMAND_ID,
        label: 'Reload Window',
      },
      order: 2,
      group: '2_views',
    });
  }

  registerKeybindings(registry: KeybindingRegistry) {
    registry.unregisterKeybinding({
      command: 'electron.reload',
      keybinding: 'shift+ctrlcmd+r',
    });
    registry.registerKeybinding({
      command: COMMAND_ID,
      keybinding: 'shift+ctrlcmd+d',
    });
  }
}

```

**Comment by bytemain** [comment]
可以打开 快捷键面板看看快捷键是否真的设置了：
<img width="217" alt="CleanShot 2022-10-09 at 15 56 09@2x" src="https://user-images.githubusercontent.com/13938334/194744857-faae1c70-cc41-48ca-a26e-0349b05217ba.png">
确认一下相关注册 keybinding 的代码是否执行了

**Comment by bytemain** [comment]
注意应用侧的 Module 需要在 `@Domain` 里加上一个KeybindingContribution。

![image](https://user-images.githubusercontent.com/13938334/194746304-f3cb5711-2271-441d-a776-f90e7b7754e6.png)


这个 bydesign 的逻辑，贡献点必须声明了才会生效。

因为模块是通过 @Domain 这个装饰器来声明自己有哪些 Contribution，然后框架内有函数来获取哪些模块是有 KeybindingContribution 的，再执行这些模块。

**Comment by bytemain** [comment, marked as answer]
有关贡献点机制可以先读一下我们的官方文档：https://opensumi.com/zh/docs/develop/basic-design/contribution-point
